[futures] Add torch.futures.collect_all()/wait_all() python api.

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39790 [futures] Add torch.futures.collect_all()/wait_all() python api.**

The "[fut.wait() for fut in futs]" idiom can introduce up to
O(len(futs)) thread switches, which may be excessive for large N.

This plumbs through the new c++ c10::collectAll() to Python space
so that we only employ a single python-side wait.

Differential Revision: [D21976891](https://our.internmc.facebook.com/intern/diff/D21976891/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21976891/)!

Commit:
[WIP] Add rpc.collect_all()/wait_all() to rpc python api.

The "[fut.wait() for fut in futs]" idiom can introduce up to
O(len(futs)) thread switches, which may be excessive for large N.

This plumbs through the new c++ c10::collectAll() to Python space
so that we only employ a single jit-side wait.

Differential Revision: [D21976891](https://our.internmc.facebook.com/intern/diff/D21976891/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21976891/)!

[ghstack-poisoned]

Commit:
Update on "[WIP] Add rpc.collect_all()/wait_all() to rpc python api."


The "[fut.wait() for fut in futs]" idiom can introduce up to
O(len(futs)) thread switches, which may be excessive for large N.

This plumbs through the new c++ c10::collectAll() to Python space
so that we only employ a single jit-side wait.

Differential Revision: [D21976891](https://our.internmc.facebook.com/intern/diff/D21976891/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21976891/)!

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -768,3 +769,29 @@ def rpc_async(to, func, args=None, kwargs=None, timeout=UNSET_RPC_TIMEOUT):
     """
     return _invoke_rpc(to, func, RPCExecMode.ASYNC, args, kwargs, timeout)
+
+def collect_all(futures):
```
> Do we expect torch.futures to be used anywhere else than rpc?

Yes, we could replace `ProcessGroup::work` with that future. And there was also a discussion to let `loss.backward()` to return a `Future`. If that occurs, most likely it will return `torch.futures.Future`.

kiukchung replied:
Gotcha, yea in this case we should have these be in `torch.futures` so that the user doesn't have to import `torch.rpc` only to get access to `wait|collect_all()`

jjlilley replied:
Makes sense in retrospect, just moved to torch.futures package

kiukchung replied:
@jjlilley are you planning to remove `wait|collect_all()` from `rpc`? I see that you've moved the two functions to `futures` but for some reason I don't see them removed from `rpc`...

jjlilley replied:
Sorry about that! I actually removed them in my buffer, but my "C-x C-s" apparently didn't go through. Uploading a new version

kiukchung replied:
ah that emacs... :)

Commit:
move collect_all to torch.futures. on "[WIP] Add rpc.collect_all()/wait_all() to rpc python api."


The "[fut.wait() for fut in futs]" idiom can introduce up to
O(len(futs)) thread switches, which may be excessive for large N.

This plumbs through the new c++ c10::collectAll() to Python space
so that we only employ a single jit-side wait.

Differential Revision: [D21976891](https://our.internmc.facebook.com/intern/diff/D21976891/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21976891/)!

[ghstack-poisoned]

Commit:
rebase on "[WIP] Add rpc.collect_all()/wait_all() to rpc python api."


The "[fut.wait() for fut in futs]" idiom can introduce up to
O(len(futs)) thread switches, which may be excessive for large N.

This plumbs through the new c++ c10::collectAll() to Python space
so that we only employ a single jit-side wait.

Differential Revision: [D21976891](https://our.internmc.facebook.com/intern/diff/D21976891/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21976891/)!

[ghstack-poisoned]

Commit:
Update on "[futures] Add torch.futures.collect_all()/wait_all() python api. "


The "[fut.wait() for fut in futs]" idiom can introduce up to
O(len(futs)) thread switches, which may be excessive for large N.

This plumbs through the new c++ c10::collectAll() to Python space
so that we only employ a single python-side wait.

Differential Revision: [D21976891](https://our.internmc.facebook.com/intern/diff/D21976891/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21976891/)!

[ghstack-poisoned]

Commit:
Update on "[futures] Add torch.futures.collect_all()/wait_all() python api. "


The "[fut.wait() for fut in futs]" idiom can introduce up to
O(len(futs)) thread switches, which may be excessive for large N.

This plumbs through the new c++ c10::collectAll() to Python space
so that we only employ a single python-side wait.

Differential Revision: [D21976891](https://our.internmc.facebook.com/intern/diff/D21976891/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21976891/)!

[ghstack-poisoned]

mruberry commented:
Reverting. This appears to have broken Windows builds, as the OSS CI on this PR shows. 

Sample log from pytorch_windows_vs2017_14.16_py36_cpu_build:

```
caffe2\CMakeFiles\torch_cpu.dir\__\torch\csrc\api\src\serialize\output-archive.cpp.obj win_tmp\mkl\lib\libiomp5md.lib win_tmp\mkl\lib\libiomp5md.lib LIB\cpuinfo.lib LIB\fbgemm.lib LIB\foxi_loader.lib win_tmp\mkl\lib\mkl_intel_lp64.lib win_tmp\mkl\lib\mkl_intel_thread.lib win_tmp\mkl\lib\mkl_core.lib win_tmp\mkl\lib\libiomp5md.lib win_tmp\mkl\lib\mkl_intel_lp64.lib win_tmp\mkl\lib\mkl_intel_thread.lib win_tmp\mkl\lib\mkl_core.lib win_tmp\mkl\lib\libiomp5md.lib win_tmp\mkl\lib\mkl_intel_lp64.lib win_tmp\mkl\lib\mkl_intel_thread.lib win_tmp\mkl\lib\mkl_core.lib win_tmp\mkl\lib\libiomp5md.lib LIB\dnnl.lib LIB\cpuinfo.lib -WHOLEARCHIVE:C:/Users/circleci/project/build/LIB/caffe2_protos.lib LIB\clog.lib LIB\asmjit.lib -WHOLEARCHIVE:C:/Users/circleci/project/build/LIB/onnx.lib LIB\onnx_proto.lib LIB\libprotobuf.lib -WHOLEARCHIVE:C:/Users/circleci/project/build/LIB/Caffe2_perfkernels_avx.lib -WHOLEARCHIVE:C:/Users/circleci/project/build/LIB/Caffe2_perfkernels_avx2.lib LIB\c10.lib -WHOLEARCHIVE:C:/Users/circleci/project/build/LIB/Caffe2_perfkernels_avx512.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib 
   Creating library LIB\torch_cpu.lib and object LIB\torch_cpu.exp
ninja: build stopped: subcommand failed.
-- Building version 1.6.0a0+1409937
cmake -GNinja -DBUILD_ENVIRONMENT=pytorch-win-vs2017-14-16-cpu-py3 -DBUILD_PYTHON=True -DBUILD_TEST=True -DBUILD_TYPE=release -DCMAKE_BUILD_TYPE=Release -DCMAKE_GENERATOR=Ninja -DCMAKE_INCLUDE_PATH=C:\Users\circleci\project\build\win_tmp\mkl\include -DCMAKE_INSTALL_PREFIX=C:\Users\circleci\project\torch -DCMAKE_PREFIX_PATH=C:\Jenkins\Miniconda3\Lib\site-packages -DCMAKE_VERBOSE_MAKEFILE=1 -DCUDA_NVCC_EXECUTABLE=C:\Users\circleci\project\build\win_tmp\bin\randomtemp.exe -DJAVA_HOME=C:\Program Files\OpenJDK\jdk-12.0.2 -DNUMPY_INCLUDE_DIR=C:\Jenkins\Miniconda3\lib\site-packages\numpy\core\include -DPYTHON_EXECUTABLE=C:\Jenkins\Miniconda3\python.exe -DPYTHON_INCLUDE_DIR=C:\Jenkins\Miniconda3\include -DPYTHON_LIBRARY=C:\Jenkins\Miniconda3/libs/python36.lib -DTORCH_BUILD_VERSION=1.6.0a0+1409937 -DUSE_CUDA=0 -DUSE_NUMPY=True C:\Users\circleci\project
cmake --build . --target install --config Release -- -j 16
Traceback (most recent call last):
  File "setup.py", line 732, in <module>
    build_deps()
  File "setup.py", line 316, in build_deps
    cmake=cmake)
  File "C:\Users\circleci\project\tools\build_pytorch_libs.py", line 62, in build_caffe2
    cmake.build(my_env)
  File "C:\Users\circleci\project\tools\setup_helpers\cmake.py", line 345, in build
    self.run(build_args, my_env)
  File "C:\Users\circleci\project\tools\setup_helpers\cmake.py", line 141, in run
    check_call(command, cwd=self.build_dir, env=env)
  File "C:\Jenkins\Miniconda3\lib\subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--target', 'install', '--config', 'Release', '--', '-j', '16']' returned non-zero exit status 1.
+ cleanup
+ retcode=1
+ set +x
```

jjlilley commented:
fwiw, we just needed the c10:: namespace prefix in a couple references to build on Windows, follow-up change: https://github.com/pytorch/pytorch/pull/39964
